### PR TITLE
makes guest API more similar to host

### PR DESCRIPTION
**Pull request by codefromthecrypt**
#### What type of PR is this?

/kind cleanup
/kind api-change

#### What this PR does / why we need it:

This makes the parameters to filter function more like those in the host by creating a proxy interface. Doing so only slows down the no-op case by some tens of nanoseconds, so it doesn't matter.

The interesting part is if users return nil for status instead of multiple values it is faster. So, this changes signature to return a pointer to status. (I checked against stack based value and it is still faster).

Finally, this optimizes the example, because our test data doesn't actually have podSpecNodeName field. This drop in execution time proves parameters not read are lazy.

#### Does this PR introduce a user-facing change?

NONE because the project isn't usable yet.

#### What are the benchmark results of this change?

First, changing to new guest signatures
```
goos: darwin
goarch: arm64
pkg: sigs.k8s.io/kube-scheduler-wasm-extension/internal/e2e
                                            │   old.txt   │              new.txt               │
                                            │   sec/op    │   sec/op     vs base               │
PluginFilter/noop/params:_small-12            147.2n ± 1%   187.8n ± 0%  +27.54% (p=0.002 n=6)
PluginFilter/noop/params:_real-12             148.4n ± 1%   186.4n ± 1%  +25.57% (p=0.002 n=6)
PluginFilter/filter-simple/params:_small-12   7.274µ ± 1%   6.518µ ± 6%  -10.39% (p=0.002 n=6)
PluginFilter/filter-simple/params:_real-12    113.9µ ± 1%   101.2µ ± 1%  -11.21% (p=0.002 n=6)
geomean                                       2.063µ        2.192µ        +6.25%
```

Next, optimizing the filter example
```
goos: darwin
goarch: arm64
pkg: sigs.k8s.io/kube-scheduler-wasm-extension/internal/e2e
                                            │   old.txt    │              new.txt               │
                                            │    sec/op    │   sec/op     vs base               │
goos: darwin
goarch: arm64
pkg: sigs.k8s.io/kube-scheduler-wasm-extension/internal/e2e
                                            │   old.txt    │              new.txt               │
                                            │    sec/op    │   sec/op     vs base               │
PluginFilter/noop/params:_small-12             147.2n ± 1%   187.3n ± 2%  +27.20% (p=0.002 n=6)
PluginFilter/noop/params:_real-12              148.4n ± 1%   187.1n ± 3%  +26.08% (p=0.002 n=6)
PluginFilter/filter-simple/params:_small-12    7.274µ ± 1%   6.564µ ± 3%   -9.75% (p=0.002 n=6)
PluginFilter/filter-simple/params:_real-12    113.95µ ± 1%   49.28µ ± 2%  -56.75% (p=0.002 n=6)
geomean                                        2.063µ        1.835µ       -11.05%
```

### CODE REVIEW

File: Makefile
Comment (codefromthecrypt) [inline Makefile:8]: build was broke

**Comment by codefromthecrypt** [comment]
ps I ran tests manually, and they pass. after https://github.com/kubernetes-sigs/kube-scheduler-wasm-extension/pull/30 we don't need manual verification

File: guest/filter.go
Comment (anuraaga) [inline guest/filter.go:72]: Minor but I think using non-naked return makes this block easier to grok at the cost of just 5 letters

File: guest/filter.go
Comment (anuraaga) [inline guest/filter.go:104]: Ditto

**Review by anuraaga — COMMENTED** [review]
As may be noticed by now I've started watching this repo and figured can leave some comments. I can stop though if it's annoying 😄

**Review by sanposhiho — COMMENTED** [review]
one nit

**Comment by codefromthecrypt** [comment]
@anuraaga your review is helpful thanks! many things I'm not committed to, and it is helpful to see someone looking!

File: guest/filter.go
Comment (codefromthecrypt) [reply guest/filter.go:72]: agree!

**Comment by codefromthecrypt** [comment]
I should disclose the simple wasm filter grew 9KB over this, but it is 9/876 and probably unanimously worth it

File: guest/internal/imports/host.go
Comment (codefromthecrypt) [inline guest/internal/imports/host.go:27]: refactored this out so I can use it in `ScorePlugin` without copy/pasta

**Comment by codefromthecrypt** [comment]
@anuraaga can you take another look and approve or let me know if any glitches left? since this has no formal approvals to lose yet, I snuck in some more godoc and polish.

File: guest/filter.go
Comment (codefromthecrypt) [inline guest/filter.go:98]: later I want to make this generic, but not sure you can make a generic function of a pointer...

**Review by anuraaga — APPROVED** [review]
Thanks!

File: guest/filter/filter.go
Comment (codefromthecrypt) [inline guest/filter/filter.go:38]: putting this in a separate package allows the guest developer to opt-into what it is using in a way the host can see.

File: scheduler/plugin/guest.go
Comment (codefromthecrypt) [inline scheduler/plugin/guest.go:108]: as we add new plugins, we add detection logic here. Some plugins might be composite because the developer wants to implement both filter and score for better performance vs defining multiple plugins.

File: scheduler/plugin/plugin.go
Comment (codefromthecrypt) [inline scheduler/plugin/plugin.go:55]: This is how we support current day plugin use, where a type assertion happens on the result of a pluginFactory.

as we support more plugin types, this logic will get more complicated, a cartesian product perhaps of many combinations. However, the cardinality of plugin types is fixed.

File: scheduler/plugin/plugin_test.go
Comment (codefromthecrypt) [inline scheduler/plugin/plugin_test.go:200]: now we show how something that is wasm, but not a plugin shows up. Basically it is an error not export support for at least one plugin type.

**Review by codefromthecrypt — COMMENTED** [review]
I revised for work-in-progress score code, since no approvals yet. Notably I found out that the plugin system today relies on type assertions and we didn't handle that. Now we do.

File: scheduler/plugin/plugin.go
Comment (kerthcet) [inline scheduler/plugin/plugin.go:81]: Shouldn't we close runtime here?

**Review by kerthcet — COMMENTED** [review]
one question, then I can approve, cc @sanposhiho for more advices.

File: scheduler/plugin/plugin.go
Comment (codefromthecrypt) [reply scheduler/plugin/plugin.go:81]: yep

Comment (codefromthecrypt) [reply scheduler/plugin/plugin.go:91]: there was actually another, too. I refactored the code to make leaks less easy.

**Review by sanposhiho — APPROVED** [review]
/lgtm

several good ideas here, thanks

**Pull request merged**
